feat(389): Prevent concurrent authentication using a lockfile.

eric-poitras opened the pull request:
Here is my shot at implementing feature #389. Reviews and feedback are welcome!

This PR implements a mutex around the authentication process to prevent the starvation of available local port to perform the authentication flow. Also, for the user experience, it might not be desirable to start multiple authentication processes in parallel.

This implementation have the following feature:
- Use exclusive access lock on a lock file to create cross-process mutex.
- Use the file `$TMPDIR/.kubelogin.get-token.lock`
- Should work on all the major OSs.

And limitations:
- It is dependent on the configuration of the temporary folder.
- Not very customizable. (Lock file location, name)
- Only protect the credential plugin command.

int128 commented on pkg/adaptors/mutex/mutex.go:
```diff
@@ -0,0 +1,95 @@
+		return nil, ctx.Err()
+	case err := <-lockChan:
+		if err != nil {
+			err = xerrors.Errorf("error acquiring lock on file %s: %w", lfn, err)
+			closeAndDelete(fm, lfn, m.Logger)
```
Do you have some reason to return `Lock` value even if error occurred? I think the following code is better for understanding the error behavior.

```suggestion
			closeAndDelete(fm, lfn, m.Logger)
			return nil, xerrors.Errorf("error acquiring lock on file %s: %w", lfn, err)
```

eric-poitras replied:
No reason. I fixed it.

int128 commented:
Thank you for your contribution! I've tested this feature on macos and works well. [`go-filemutexz`](https://github.com/alexflint/go-filemutex) supports windows as well and looks good.

int128 commented:
I rebased your commit to run the tests. Could you check the failed test?

```
=== RUN   TestMutex/Test_successful_parallel_acquisition_with_no_reentry_allowed
--- PASS: TestMutex (0.00s)
    --- PASS: TestMutex/Test_successful_parallel_acquisition_with_no_reentry_allowed (0.00s)
panic: send on closed channel

goroutine 22 [running]:
github.com/int128/kubelogin/pkg/adaptors/mutex.TestMutex.func1.1()
	/home/circleci/project/pkg/adaptors/mutex/mutex_test.go:37 +0x1ff
created by github.com/int128/kubelogin/pkg/adaptors/mutex.TestMutex.func1
	/home/circleci/project/pkg/adaptors/mutex/mutex_test.go:42 +0x278
FAIL	github.com/int128/kubelogin/pkg/adaptors/mutex	0.091s
```

eric-poitras commented:
> I rebased your commit to run the tests. Could you check the failed test?
> 
> ```
> === RUN   TestMutex/Test_successful_parallel_acquisition_with_no_reentry_allowed
> --- PASS: TestMutex (0.00s)
>     --- PASS: TestMutex/Test_successful_parallel_acquisition_with_no_reentry_allowed (0.00s)
> panic: send on closed channel
> 
> goroutine 22 [running]:
> github.com/int128/kubelogin/pkg/adaptors/mutex.TestMutex.func1.1()
> 	/home/circleci/project/pkg/adaptors/mutex/mutex_test.go:37 +0x1ff
> created by github.com/int128/kubelogin/pkg/adaptors/mutex.TestMutex.func1
> 	/home/circleci/project/pkg/adaptors/mutex/mutex_test.go:42 +0x278
> FAIL	github.com/int128/kubelogin/pkg/adaptors/mutex	0.091s
> ```

Sure will do! 👀

eric-poitras commented:
Hi sir. Here, I fixed a couple of issues:

- I removed the code that attempted to remove the file after a lock release since it caused race conditions in the file system under load. I am not concerned too much about the temporary file since there is a single file ever created in the actual version.
- Learned about the '-race' option on test and saw that it did not like that I called the WaitGroup.Add from within the go-routine. So moved the call to Add outside.
- Learned about the golangci-lint, installed it and fixed the reported issues.
- Fixed code in relation to your review. Thanks !

I think everything is good but tell me if you need anything else.

Have a nice day!

int128 commented:
I will take a look at the conflict with another PR:

```
    --- FAIL: TestGetToken_Do/MultiUserOption (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xd47be9]

goroutine 23 [running]:
testing.tRunner.func1.1(0xdd57a0, 0x1365120)
	/usr/local/go/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc000183800)
	/usr/local/go/src/testing/testing.go:1075 +0x636
panic(0xdd57a0, 0x1365120)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/golang/mock/gomock.(*Controller).Finish(0xc0002418f0)
	/home/circleci/go/pkg/mod/github.com/golang/mock@v1.4.4/gomock/controller.go:246 +0x3b2
panic(0xdd57a0, 0x1365120)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/int128/kubelogin/pkg/usecases/credentialplugin.(*GetToken).Do(0xc0001b7b08, 0xf7a240, 0xc0001a6018, 0xe8ee8f, 0x1b, 0xe87ab8, 0xe, 0xe899da, 0x12, 0x0, ...)
	/home/circleci/project/pkg/usecases/credentialplugin/get_token.go:57 +0x109
github.com/int128/kubelogin/pkg/usecases/credentialplugin.TestGetToken_Do.func3(0xc000183800)
	/home/circleci/project/pkg/usecases/credentialplugin/get_token_test.go:265 +0x1a58
testing.tRunner(0xc000183800, 0xc000241770)
	/usr/local/go/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x5bc
FAIL	github.com/int128/kubelogin/pkg/usecases/credentialplugin	0.183s
```

binboum commented:
Hi, Would there not be an impact when we are several users on a machine to consume the service?

int128 commented:
If [`os.TempDir()`](https://golang.org/pkg/os/#TempDir) returns `/tmp`, it may conflict with another user. This feature is to prevent conflict of the local server port, so it would be nice to include the port number to lock filename.
